lib/logstash/inputs/imap.rb

```diff
@@ -186,32 +228,35 @@ def parse_mail(mail)
       # Add fields: Add message.header_fields { |h| h.name=> h.value }
       mail.header_fields.each do |header|
         # 'header.name' can sometimes be a Mail::Multibyte::Chars, get it in String form
-        name = @lowercase_headers ? header.name.to_s.downcase : header.name.to_s
+        name = header.name.to_s
+
+        # Assume we already processed the 'date' above.
+        next if name == "Date"
+
+        name = name.downcase if @lowercase_headers
+
         # Call .decoded on the header in case it's in encoded-word form.
         # Details at:
         #   https://github.com/mikel/mail/blob/master/README.md#encodings
         #   http://tools.ietf.org/html/rfc2047#section-2
         value = transcode_to_utf8(header.decoded.to_s)
 
-        # Assume we already processed the 'date' above.
-        next if name == "Date"
-
-        case (field = event.get(name))
+        targeted_name = "#{@headers_target}[#{name}]"
```

Comment:
Example: pre-normalizing the `@headers_target` isn't strictly necessary, as the syntax allows for [Composite References](https://www.elastic.co/guide/en/logstash/master/field-references-deepdive.html#formal-grammar-event-api-composite-field-reference)

```suggestion
        targeted_name = "[#{@headers_target}][#{name}]"
```

Reply:
thanks, would still rather keep the 'canonical' form here - subjectively find it less confusing than the composite.
we end up using what the user set with the exception of the bare `foo` form being wrapped (`[foo]`), the normalization code is similar to what we have in the field reference validator.

Reply:
this brings up a good question though, when user explicitly sets a `headers_target => ''`
think we should treat it as a way to disabled the target -> user does not want headers (already present in #56 but I think it makes sense to ship here) instead of treating it as if `headers_target` was not set and setting stuff at top level.

or does this look confusing?

Reply:
I agree that setting `headers_target => ''` should read as an explicit instruction to not place the headers anywhere, and distinctly different than _not_ supplying a `headers_target` instruction. We will need to call this out in the docs, and will need to add specs to ensure that it passes validation.

When the `email` RFC gets finalized, we will need a separate way to opt into the RFC format, and when this future mode is active we can also inject better defaults for related options.

Reply:
> I agree that setting `headers_target => ''` should read as an explicit instruction to not place the headers anywhere, and distinctly different than not supplying a headers_target instruction. We will need to call this out in the docs, and will need to add specs to ensure that it passes validation.

:heavy_check_mark: (pending 6.x compatibility `headers_target => ''` https://github.com/logstash-plugins/logstash-mixin-validator_support/pull/2)

there's one extra change since last review - not skipping the `date` header (despite the header being duplicate information as it's also used for `event.timestamp = LogStash::Timestamp.new(mail.date.to_time)`) - initially considered it a bug but the `date` header was present with the plugins defaults due: 
- `lowercase_headers` being on by default
- the plugin was checking the header name [to == `"Date"`](https://github.com/logstash-plugins/logstash-input-imap/blob/v3.1.0/lib/logstash/inputs/imap.rb#L197) after the lower-case operation

so while it looks like a bug the "fix" (to properly skip the header regardless or `lowercase_headers` setting) might break pipelines thus an alternate resolution to never skip the date headers seems more appropriate.